### DISCUSSION TITLE
How to retrieve the PGN from the edit mode viewer

### DISCUSSION THREAD

**Discussion by hecodes** [mliebelt/pgn-viewer#371] [Q&A]
Hi there - Great work first! 

I want to use the edit mode PgnViewerJS in my react app so that I can save the PGN string into a variable and later on upload to a website through REST API.

Can you give an example on how to access the PGN file/string after user input the moves?

**Comment by mliebelt** [comment]
Good question, I have to look that up. Hopefully there is something that can already used (if not being obvious). If not, it is a natural extension of the API I want to build since ... forever. But that is the next thing on my agenda, to have a (simple) UI to use the player/editor by sending calls.

**Comment by hecodes** [comment]
Thank you Markus!

**Reply by mliebelt** [reply]
Ok, it is a hack, but anyway ...

You have to do the following steps, to see it working:

1. You have an HTML file where something like the following is called: `let pe = PGNV.pgnEdit("board", {})`.
2. So you have access to the viewer after it is initialized. 
3. Then the following expression returns the current state of the pgn: `pe.base.getPgn().writePgn()`

What is the meaning of that?

* `pe`: This returns a JSON that contains the so called base object from the call to pgnEdit.
* `pe.base`: the base value of the JSON returned ==> the viewer.
* `pe.base.getPgn()`: returns the so called pgnReader, the inner model (no UI).
* `pe.base.getPgn().writePgn()`: writes out the current pgn (held by the reader) as a string. If you have not entered anything, this will be empty, later it will contain all lines. At the moment, there is no functionality to get only the base line, ...

**Comment by hecodes** [comment]
Thanks!

Is there more specific instructions for ReactJS?

**Reply by mliebelt** [reply]
There is https://github.com/mliebelt/PgnViewerJS/blob/main/react.md, which explains in a nutshell how to include pgn-viewer in ReactJS. I do not know if you are aware of that. Besides that, I don't have any information. But the point in the example source code is, that you have to have the reference to the base instance somehow, which is not included in the code. See line 16 of that file, where the call to `pgnView` is done.

**Comment by coffeeDev98** [comment]
@hecodes 
I've found a way to do it in React... might be a bit hacky
You can import pgnBase from pgn-viewer library and recreate the pgnEdit function within your file

```
const pgnEdit = function (boardId: any, configuration: any) {
    let base = pgnBase(
      boardId,
      Object.assign(
        {
          showFen: true,
          mode: "edit",
          movable: {
            free: false,
            events: {
              after: function (orig: any, dest: any, meta: any) {
                base.onSnapEnd(orig, dest, meta);
                console.log(base.getPgn().writePgn()); //accessing the PGN data as mentioned by @mliebelt 
              },
            },
          },
          viewOnly: false,
        },
        configuration
      )
    );
    base.generateHTML();
    base.generateMoves();
    let board = base.generateBoard();
    return { base, board };
  };
```

PS: You cannot store the PGN in hooks or state variables.... It causes the UI to crash or not mount for some reason.

You can however store it in an Object and define a property like below

```
  const lastStats: any = {}; //must be an Object
  let lastStatsPgn = "";

  Object.defineProperty(lastStats, "pgn", {
    get() {
      return lastStatsPgn;
    },
    set(val) {
      onLastStatsValueChange(lastStatsPgn, val); //Call function to be executed whenever PGN changes
      lastStatsPgn = val; //Set new value of PGN
    },
  });
```

**Reply by hecodes** [reply]
Dear @coffeeDev98 - this is great! Can you wrap everything inside a React function component and a simple working example? much appreciated!

**Reply by coffeeDev98** [reply]
Using the example from the [React.md](https://github.com/mliebelt/PgnViewerJS/blob/main/react.md) file
```
import React, {useLayoutEffect} from 'react'
import Children from 'react-children-utilities'
import * as uuid from 'uuid'
import { pgnBase } from '@mliebelt/pgn-viewer'

const pgnEdit = function (boardId: any, configuration: any) {
    let base = pgnBase(
      boardId,
      Object.assign(
        {
          showFen: true,
          mode: "edit",
          movable: {
            free: false,
            events: {
              after: function (orig: any, dest: any, meta: any) {
                base.onSnapEnd(orig, dest, meta);
                console.log(base.getPgn().writePgn()); //accessing the PGN data as mentioned by @mliebelt 
              },
            },
          },
          viewOnly: false,
        },
        configuration
      )
    );
    base.generateHTML();
    base.generateMoves();
    let board = base.generateBoard();
    return { base, board };
  };

function PGNViewer(props) {
  const gameDecription = Children.onlyText(props.children)
  const id = 'board-' + uuid.v4()

  useLayoutEffect(() => {
    pgnEdit(id,
      {
        pgn: gameDecription,
        timerTime: '1',
        locale: 'pl',
        startPlay: 1,
        showResult: true,
        boardSize: '340',
        showFen: true,
        pieceStyle: 'merida'
      }
    )
  })

  return (
    <div id={id}></div>
  )
}

// Usage

<PGNViewer>
[Event "F/S Return Match"]
[Site "Belgrade"]
[Date "1992.11.04"]
[Round "29"]
[White "Fischer, Robert J."]
[Black "Spassky, Boris V."]
[Result "1/2-1/2"]

1. e4 e5 2. Nf3 Nc6 3. Bb5 a6 4. Ba4 Nf6 5. O-O Be7 6. Re1 b5 7. Bb3 d6 8. c3 O-O 9. h3 Nb8  10. d4 Nbd7
11. c4 c6 12. cxb5 axb5 13. Nc3 Bb7 14. Bg5 b4 15. Nb1 h6 16. Bh4 c5 17. dxe5 Nxe4 18. Bxe7 Qxe7 19. exd6 Qf6 
20. Nbd2 Nxd6 21. Nc4 Nxc4 22. Bxc4 Nb6 23. Ne5 Rae8 24. Bxf7+ Rxf7 25. Nxf7 Rxe1+ 26. Qxe1 Kxf7 27. Qe3 Qg5 
28. Qxg5 hxg5 29. b3 Ke6 30. a3 Kd6 31. axb4 cxb4 32. Ra5 Nd5 33. f3 Bc8 34. Kf2 Bf5 35. Ra7 g6 36. Ra6+ Kc5 
37. Ke1 Nf4 38. g3 Nxh3 39. Kd2 Kb5 40. Rd6 Kc5 41. Ra6 Nf2 42. g4 Bd3 43. Re6 1/2-1/2
</PGNViewer>
```

Should Ideally work. Haven't tested this particular snippet. I have used a similar one in my project though.

**Comment by hecodes** [comment]
Thanks, I'll check it out.
